timemachine/exchange/exchange_mover.py

```diff
@@ -0,0 +1,354 @@
+
+    host_bps = []
+    for p, bp in zip(params, potentials):
+        host_bps.append(bp.bind(p))
+
+    temperature = DEFAULT_TEMP
+    dt = 1e-3
+
+    integrator = LangevinIntegrator(temperature, dt, 1.0, combined_masses, seed)
+
+    bond_list = get_bond_list(host_bps[0].potential)
+    group_idxs = get_group_indices(bond_list, len(combined_masses))
+    barostat_interval = 25
+
+    barostat = MonteCarloBarostat(
+        len(combined_masses), DEFAULT_PRESSURE, temperature, group_idxs, barostat_interval, seed + 1
+    )
+
+    # (YTZ): This is disabled because the initial ligand and starting waters are pre-minimized
+    # print("Minimizing the host system...", end="", flush=True)
+    # host_conf = minimize_host_4d([mol], host_config, ff)
+    # print("Done", flush=True)
+
+    host_conf = solvent_conf
+    combined_conf = np.concatenate([host_conf, get_romol_conf(mol)], axis=0)
+
+    ligand_idxs = np.arange(num_water_atoms, num_water_atoms + mol.GetNumAtoms())
+
+    initial_state = InitialState(
+        host_bps, integrator, barostat, combined_conf, np.zeros_like(combined_conf), solvent_box, 0.0, ligand_idxs
+    )
+
+    num_water_mols = num_water_atoms // 3
+
+    return initial_state, num_water_mols, solvent_topology
+
+
+class ExchangeMover:
+    def __init__(self, nb_params, nb_beta, nb_cutoff, water_idxs):
+        self.nb_beta = nb_beta
+        self.nb_cutoff = nb_cutoff
+        self.nb_params = jnp.array(nb_params)
+        self.num_waters = len(water_idxs)
+        self.water_idxs = water_idxs
+        # counters
+        self.a_count = 0  # accept count
+        self.t_count = 0  # trial count
+        self.beta = 1 / DEFAULT_KT
+
+        @jax.jit
+        def U_fn(conf, box, a_idxs, b_idxs):
+            # compute the energy of an interaction group
+            conf_i = conf[a_idxs]
+            conf_j = conf[b_idxs]
+            params_i = self.nb_params[a_idxs]
+            params_j = self.nb_params[b_idxs]
+            return nonbonded.nonbonded_block(conf_i, conf_j, box, params_i, params_j, self.nb_beta, self.nb_cutoff)
+
+        self.U_fn = U_fn
+
+    def exchange_move(self, coords, box):
+        self.t_count += 1
+        n_atoms = len(coords)
+        chosen_water = np.random.randint(self.num_waters)
+        chosen_water_atoms = self.water_idxs[chosen_water]
+
+        # compute delta_U of deletion
+        a_idxs = chosen_water_atoms
+        b_idxs = np.delete(np.arange(n_atoms), a_idxs)
+
+        # compute delta_U of insertion
+        trial_chosen_coords = coords[chosen_water_atoms]
+        trial_translation = np.diag(box) * np.random.rand(3)
+        # tbd - what should we do  with velocities?
+
+        # this has a higher acceptance probability than if we allowed for rotations
+        # (probably because we have a much higher chance of a useless move)
+        moved_coords = randomly_translate(trial_chosen_coords, trial_translation)
```

Comment:
Good observation.

(Suggests that biasing the distribution of proposed translations away from (0,0,0) might be a useful optimization in the future.)

Reply:
Yep - though we'd want to quantify also the productivity of the accepted moves, i.e. a change in the occupancy